### ISSUE TITLE
K-9 is unable to fetch more than 9 messages from an Exchange server

### ISSUE THREAD

**Issue by philipwhiuk** [thunderbird/thunderbird-android#1143]
## Expected behaviour

Email headers (and content) should be retrieved
### Actual behaviour

IndexOutOfBoundsException

```
03-05 01:02:12.967    5668-5723/? I/k9﹕ Synchronizing folder Work:Inbox
03-05 01:02:12.974    5668-5723/? I/k9﹕ url = <<REMOVED FOR PRIVACY>>
03-05 01:02:12.974    5668-5723/? I/k9﹕ Requesting gzipped data
03-05 01:02:13.111    5668-5723/? I/k9﹕ url = <<REMOVED FOR PRIVACY>>
03-05 01:02:13.111    5668-5723/? I/k9﹕ Requesting gzipped data
03-05 01:02:13.359    5668-5723/? E/k9﹕ synchronizeMailbox
    java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
            at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
            at java.util.ArrayList.set(ArrayList.java:481)
            at com.fsck.k9.mail.store.WebDavStore$WebDavFolder.fetchEnvelope(WebDavStore.java:1735)
            at com.fsck.k9.mail.store.WebDavStore$WebDavFolder.fetch(WebDavStore.java:1512)
            at com.fsck.k9.controller.MessagingController.fetchUnsyncedMessages(MessagingController.java:1475)
            at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1312)
            at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:1124)
            at com.fsck.k9.controller.MessagingController.access$300(MessagingController.java:113)
            at com.fsck.k9.controller.MessagingController$8.run(MessagingController.java:919)
            at com.fsck.k9.controller.MessagingController.run(MessagingController.java:423)
            at java.lang.Thread.run(Thread.java:818)
03-05 01:02:13.412    5668-5723/? E/k9﹕ Failed synchronizing folder Work:Inbox @ Sat Mar 05 01:02:13 GMT+00:00 2016
```
### Steps to reproduce
1. Add an Exchange account
2. Sync emails
### Environment

K-9 Mail version: 5.107, master

Android version: 6.0.1

Account type (IMAP, POP3, WebDAV/Exchange): Exchange

**Comment by philipwhiuk** [comment]
The bug is with:

```
            List<WebDavMessage> messages = new ArrayList<WebDavMessage>(10);
            // Some lines of code
            messages.set(i, startMessages.get(i));
```

[set](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html#set(int,%20E\)) replaces an existing element. But messages is guaranteed to be empty at this point - we've only just created it as an empty array.

I think we just want `message.add(startMessages.get(i))`

**Comment by philipwhiuk** [comment]
It's never as simple as it looks is it...

This appears to be the design all over the place.

A couple of lines down, the newMessages container has the same issue. Then when I fixed that on a fork, it printed out the following many times:

```
03-05 11:04:46.578 7993-8017/com.fsck.k9.debug I/k9: Requesting gzipped data
```

And then it ran into the same problem with the flags:

```
03-05 11:05:56.554 7993-8017/com.fsck.k9.debug E/k9: synchronizeMailbox
   java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
      at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
      at java.util.ArrayList.set(ArrayList.java:481)
      at com.fsck.k9.mail.store.webdav.WebDavStore$WebDavFolder.fetchFlags(WebDavStore.java:1637)
      at com.fsck.k9.mail.store.webdav.WebDavStore$WebDavFolder.fetch(WebDavStore.java:1490)
      at com.fsck.k9.controller.MessagingController.fetchUnsyncedMessages(MessagingController.java:1326)
      at com.fsck.k9.controller.MessagingController.downloadMessages(MessagingController.java:1168)
      at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:967)
      at com.fsck.k9.controller.MessagingController.access$300(MessagingController.java:99)
      at com.fsck.k9.controller.MessagingController$8.run(MessagingController.java:754)
      at com.fsck.k9.controller.MessagingController.run(MessagingController.java:257)
      at java.lang.Thread.run(Thread.java:818)

So it looks like this particular code path has never been excised at all. I'm guessing there's not much use of Exchange in K9. Certainly the IMAP store is better organised and tested.

I tried writing a test for this behaviour but I found I need to build a whole Store object which then mandated a WebDavHttpClient. I'm probably going to take a leaf out of the IMAP store design and separate out the components. This will take longer to code but be easier to test for the future.

@cketti are you okay with this kind of substantial PR?

**Comment by philipwhiuk** [comment]
Looking through the code, there's also some unimplemented methods. I'll probably try and implement those too - "Mark All Messages Read" is definitely something I want to use.

**Comment by philipwhiuk** [comment]
```
03-05 12:25:37.288 7373-7411/com.fsck.k9.debug W/SingleClientConnManager: 
   Invalid use of SingleClientConnManager: connection still allocated.
   Make sure to release the connection before allocating another one.
```

Seeing quite a bit of this too. Not sure if it's related yet.

With the fixed version (pushed to my repo), the app hung. It appeared to request more messages every time I tried to get it to poll.

I re-installed and was able to get my fixed version to work. However the status line said "Poll Work: Inbox 50/25" before it finished.

It still seems to take ages to poll, but that might be my exchange server.

**Comment by philipwhiuk** [comment]
This was fixed by #1147
